diff --git a/target/product/full_base.mk b/target/product/full_base.mk
index 7e5d450..cd6e00b 100644
--- a/target/product/full_base.mk
+++ b/target/product/full_base.mk
@@ -30,9 +30,6 @@ PRODUCT_PROPERTY_OVERRIDES := \
 # Put en_US first in the list, so make it default.
 PRODUCT_LOCALES := en_US
 
-# Include drawables for all densities
-PRODUCT_AAPT_CONFIG := normal hdpi xhdpi xxhdpi
-
 # Get some sounds
 $(call inherit-product-if-exists, frameworks/base/data/sounds/AllAudio.mk)
 
